[1주차] Database 설계 및 아키텍처 설계 구현 완료

dbdb1114 opened the pull request:
### [1주차] Database 설계 및 아키텍처 설계 구현 완료
1. 데이터베이스 설계 및 아키텍처 설계, 
2. 메인 페이지 추가
3. 메인 페이지 상영정보 조회 기능 구현 

<br/>  

### 작업 내용  
> 이번 PR에서 진행된 주요 변경 사항을 기술해 주세요.  
>**코드 구조, 핵심 로직** 등에 대해 설명해 주시면 좋습니다. (이미지 첨부 가능)  
> ex: `ConcurrentOrderService`에 동시 주문 요청 처리 기능 추가   
-  module-rds-repo : DB와의 연결과 entity, repository를 활용합니다. 
- module-service : repository를 이용하여 Data를 가져와 정렬하고, 정제하는 역할을 합니다. 
```java 
	@Override
	public List<Map.Entry<MovieDTO, List<ShowingDTO>>> getTodayShowing() {
		LocalDateTime from = LocalDateTime.now().plusMinutes(30);
		LocalDateTime to = LocalDateTime.of(LocalDate.now(), LocalTime.MAX);

		List<Showing> showingsByStTimeBetween = showingRepository.findShowingsByStTimeBetween(from, to);

		// 금일 상영 정보가 아예 없을 경우 다음날 조회
		if(showingsByStTimeBetween.isEmpty() || showingsByStTimeBetween.size() < 10){
			LocalDateTime tomorrowFrom = LocalDateTime.of(LocalDate.now().plusDays(1L), LocalTime.MIN);
			LocalDateTime tomorrowTo = to.plusDays(1L);
			showingsByStTimeBetween.addAll(showingRepository.findShowingsByStTimeBetween(tomorrowFrom, tomorrowTo));
		}

		return showingsByStTimeBetween.stream()
			.map(showingEntity -> modelMapper.map(showingEntity, ShowingDTO.class))
			.collect(Collectors.groupingBy(ShowingDTO::getMovie))
			.entrySet().stream()
			.peek(entry -> entry.getValue().sort(Comparator.comparing(ShowingDTO::getStTime)))
			.sorted(Map.Entry.comparingByKey(Comparator.comparing(MovieDTO::getOpenDay, Comparator.reverseOrder())))
			.toList();
	}
```
- module-app : module-service에서 조회한 데이터를 응답값에 맞게 수정 보완 처리하여 response를 반환합니다. 
```java
	@GetMapping("/today-info")
	public ResponseEntity<Map> getTodayShowing() {
		List<Map.Entry<MovieDTO, List<ShowingDTO>>> todayShowing = showingService.getTodayShowing();

		List<Map<String, Object>> responseData = todayShowing.stream()
			.map(entry -> Map.entry(entry.getKey(), entry.getValue().stream()
				.map(val -> modelMapper.map(val, ShowingResponse.class)).toList()))
			.map(entry -> Map.of("movie", entry.getKey(), "timeTable", entry.getValue()))
			.toList();

		if (responseData.isEmpty()) {
			return ResponseEntity.noContent().build();
		} else {
			return ResponseEntity.ok(Map.of("data", responseData));
		}
	}
```
- module-core : dto클래스를 작성했습니다. 구성은 아래와 같습니다. 
```shell
.
├── dto
│   ├── external
│   │   └── ShowingResponse.java
│   └── internal
│       ├── GenreDTO.java
│       ├── MovieDTO.java
│       ├── RatingDTO.java
│       ├── ScreenDTO.java
│       └── ShowingDTO.java
├── exception
└── util
```
### 발생했던 문제와 해결 과정을 남겨 주세요. 
-   **문제 1** - 아키텍처를 설계하면서 이정도 기능의 설계에 있어 멀티모듈 아키텍처가 어떤 긍정적인 효과를 가져올지 고민이 많았습니다. 
- **해결 방법 1** - 제가 이해하기로는 결국 멀티모듈 아키텍처를 설계하는 이유도 팀의 불편함, 개인의 불편함에서 시작된 것 같았습니다. 
그래서 제가 어떻게 하면 이전보다 나은 코드와 역할과 책임의 분리를 잘 지킬 수 있을지 고민하여 LayeredArchitecture를 기반으로 multimodule 프로젝트를 설계하게 됐습니다. 
-   **문제 2** - 레이어간 역할과 책임의 분리 어느정도 까지인지 컨트롤러의 책임과 역할은 뭐고, 서비스의 책임과 역할은 어느정도 까지인지 고민이 많았습니다.
- **해결 방법 2** - 무엇을 위해 존재하는가를 생각했습니다. 그래서 컨트롤러는 입출력을 담당하는 부분이라 생각하여, 데이터의 응답 규격에 맞추는 일을 수행하게 했고, 서비스 레이어는 비즈니스적으로 원하는 데이터를 적당히 깔끔한 규격으로 반환하게 하였습니다.

### 이번 주차에서 고민되었던 지점이나, 어려웠던 점을 알려 주세요.
> 과제를 해결하며 특히 어려웠던 점이나 고민되었던 지점이 있다면 남겨주세요. 
-  클린 코드에 대한 고민이 많았습니다. 이 메소드명은 적절한지, 뎁스는 적절한지 내 코드만 계속 보고 있으니 가독성이 좋게 느껴지는 건 아닌지 고민이 많았습니다.
-  stream api를 사용하면 데이터를 추가적인 지역변수나 제어문 선언 없이 로직을 처리할 수 있습니다. 코드가 간결해 보이긴 하지만 하다보니 중간 연산의 값을 파악하기 힘든 그런 일들이 있었습니다. 어떻게 사용하는게 좋은지 헷갈렸습니다. 
-  실무에서 어떻게 해야할지 고민이 많습니다. 저는 취준생 입장이라 현업에 나아가서 일을 하려면 지금 프로젝트를 혼자 진행하는 방식과 제 노력과 같은 부분이 적절한지 계속해서 고민했습니다. 
- 커밋 로그를 어떻게 정리해야할지 감이 잘 안 잡힙니다. service레이어를 수정하다보면 Repository 수정이 필요해지고 그러다보면 커밋이 시간흐름을 어기게 되거나 중간중간 service와 repository를 왔다 갔다 하며 난잡해지는 것 같습니다. 

### 리뷰 포인트
-  최대한 비판적으로 봐주셨으면 좋겠습니다.  고칠점을 서슴없이 말씀해주셨으면 좋겠습니다.
- stream을 과하게 사용한 것은 아닌지 리뷰 해주셨으면 좋겠습니다. 
- entity설계를 지적해주셨으면 좋겠습니다. 
- 전반적으로 보시기에 약간 거슬리는 부분들도 asis와 tobe가 명확하진 않더라도 듣고싶습니다. 

### 기타 질문
- 테스트 코드를 제대로 작성하며 프로젝트를 해본 경험이 없습니다. 참고할만한 책이나 강의나 라이브러리를 알려주시면 감사할 것 같습니다.
- 실무에서 어떻게 일하는지 궁금합니다. 예를 들어 하루에 코드를 작성하는 시간은 몇 시간이나 되는지 코드를 작성하기 위해 설계하는 시간은 얼마나 가지는지 궁금합니다. 

### 추가 
늦어서 죄송합니다. 계속 부족한 부분들이 보여 급하게 하다보니,,, 오히려 더 늦어졌습니다. 차주에는 시간을 조금 더 잘 지키도록 하겠습니다.

dbdb1114 committed:
Initial commit

dbdb1114 committed:
불필요 파일 제거

dbdb1114 committed:
[ feat ] RDS 기본 의존성 추가

spring-data-jpa. h2database, postgresql

dbdb1114 committed:
[ feat ] entity 설계 작성

추가 수정 예상되며, 테스트 혹은 개발하면서 수정 예정

dbdb1114 committed:
[ feat ] repository 작성

dbdb1114 committed:
[ fix ] 불필요 코드 제거

dbdb1114 committed:
[ refactor ] BaseEntity생성 및 상속, PK starategy 설정 추가

1. BaseEntity 작성하여 상속 하도록 수정 및 빌더 패턴 설정
2. PK Id필드 strategy = GenerationType.IDENTITY 설정 추가
3. Lombok 의존성 추가 및 빌더 패턴 도입

dbdb1114 committed:
[ fix ] screen - seats 연관관계 단방향으로 변경

dbdb1114 committed:
[ fix ]  Movie plot 필드 타입 변경

dbdb1114 committed:
[ refactor ] 칼럼명 스네이크케이스 -> 카멜케이스 수정

dbdb1114 committed:
[ feat ] test위한 util 클래스 정의

private 필드에 접근하기위한 리플렉션 유틸 메소드

dbdb1114 committed:
[ refactor ] 불필요 Import 제거

dbdb1114 committed:
[ test ] 상영정보 레파지토리 테스트 class 생성

dbdb1114 committed:
[ feat ] 상영정보 조회 메소드 작성 및 테스트

dbdb1114 committed:
[ test ] GenreRepository 테스트 클래스 생성

dbdb1114 committed:
[ test ] GenreRepository 장르 조회 테스트

dbdb1114 committed:
[ refactor ] FetchType LAZY로 변경

dbdb1114 committed:
[ test ] ShowingRepository 영화 장르 조회 테스트

dbdb1114 committed:
[ test ] ShowingRepository rating 조회 테스트 추가

dbdb1114 committed:
Merge pull request #1 from dbdb1114/feature/rds-repo

Feature/rds repo 1주차 요구사항 rds-repo 작업 완료

dbdb1114 committed:
[ conf ] ignore 항목 추가

dbdb1114 committed:
Merge branch 'dev' of https://github.com/dbdb1114/HangHaeHo into dev

dbdb1114 committed:
[ refactor ] ShowingRepository 상영정보 조회 메소드 수정

1. Orderby RDB에서 하는 게 무의미함
2. 메소드 변경에 따른 테스트 코드 수정 및 Assertions 메소드 수정

dbdb1114 committed:
[ fix ] rds 패키지 변경

테스트 코드 작성시 빈을 로드하는 context 지점을 최상위로 둘 수 있어야 하기 때문에
java.main.module 패키지 생성후 기존 패키지 모두 하위로 이동

dbdb1114 committed:
[ feat ] module-core DTO 클래스 작성

dbdb1114 committed:
[ feat ] module-service build.gradle 작성

implementation: 모듈, model-mapper
compileOnly : lombok, starter-web, data-jpa

dbdb1114 committed:
[ feat ] model mapper config 작성

dbdb1114 committed:
[ feat ] showingService 금일 상영정보 조회 기능 구현

stream 활용하여 상영정보를 영화 기준으로 groupby하고, 정해진 기준으로 정렬
( 최신 개봉을 앞으로 정렬, 상영시간은 가장 이른 시간을 앞으로 정렬 )

dbdb1114 committed:
[ test ] 상영정보 조회 메소드 테스트 코드 작성

1. 의존 주입 확인
2. groupBy 및 금일 기준 조회 여부 확인

dbdb1114 committed:
[ refactor ] Autowired -> RequiredArgsConstructor 의존 주입 방식 변경

dbdb1114 committed:
[ refactor ] service layer  상영정보 조회 메소드 반환 타입 수정

dbdb1114 committed:
[ refactor ] 패키지 변경 및 @Data 어노테이션 제거

dbdb1114 committed:
[ test ] ShowingService getTotalShowing() 수정에 따른 테스트 코드 수정

dbdb1114 committed:
[ feat ] ShowingService getTodayShowing() 메소드 기능 수정

금일 상영정보 없을 경우 분기 처리

dbdb1114 committed:
[ refactor ] 불필요 의존성 제거

dbdb1114 committed:
[ feat ] 상영정보 응답 dto 작성

dbdb1114 committed:
[ refactor ] 메인 함수 패키지 변경

dbdb1114 committed:
[ feat ] module-app ShowingController 상영정보 조회 기능 구현

dbdb1114 committed:
[ refactor ] module-app build.gradle 불필요 의존성 제거

dbdb1114 committed:
[ fix ] DTO 객체 @getter 추가

Genre,Rating,Screen
MovieEntity => MovieDTO 변환시 필요

dbdb1114 committed:
[ fix ] MovieDTO equals,hashCode 함수 추가

getTodayShowing()메소드에서 groupBy시 필요

dbdb1114 committed:
[ feat ] 메인 페이지 구현

localhost:8080/ 으로 접속시 확인 가능

dbdb1114 committed:
[ feat ] 빌드 불필요 하위 모듈 bootJar fase 처리

dbdb1114 committed:
[ fix ] ShowingService 메소드 수정에 따른 테스트 코드 수정

dbdb1114 committed:
[ feat ] ShowingService getTodayShowing() 메소드 수정

금일 상영정보가 없거나 너무 적을 경우 다음날 상영정보 추가 조회

dbdb1114 committed:
[ fix ] 프론트엔드 비동기 요청 호스트 주소 기준으로 변경

dbdb1114 committed:
[ feat ] module-rds-repo yml 파일 작성

dbdb1114 committed:
[ feat ] docker sring profiles 수정

dbdb1114 committed:
[ feat ] docker-compose 작성

dbdb1114 committed:
[ feat ] root project build.gradle 파일 수정

서브 프로젝트 임포트 정보 추가 및 공통 설정 추가

dbdb1114 committed:
[ fix ] docker-compose spring-profiles 수정 ( prod -> dev )

dbdb1114 committed:
[ feat ] ReadMe 작성

dbdb1114 committed:
[ fix ]  init sql 제거

dbdb1114 committed:
[ faet ] DDL 및 INSERT 쿼리 작성

dbdb1114 committed:
[ fix ] readme 수정

애플리케이션 실행 스크립트 수정

dbdb1114 committed:
Merge branch 'hanghae-skillup:main' into dev

dbdb1114 committed:
[ feat ] readme 작성

포크 레포지토리로 옮기면서 약간의 착오가 있어 다시 커밋합니다.

dbdb1114 committed:
[ feat [ postgresql->mysql 로 변경

dbdb1114 committed:
[ feat ] mysql 기준으로 야믈파일과 초기화 ddl파일 변경

dbdb1114 committed:
[ fix ] 칼럼명 변경 row,number -> seatRow,seatNumber

dbdb1114 committed:
[ fix ] dev yaml프로퍼티 변경

youngxpepp reviewed (commented):
안녕하세요 정현님. 이건홍 코치라고 합니다ㅎㅎ

아키텍처에 대한 고민을 많이 하신듯 합니다. 저 또한 현업에서 아키텍처에 대한 고민을 많이 하고 있고 자료를 많이 찾아봤어요. 아키텍처 선택에는 이유가 있어야 하는데 그 부분이 합리적이라고 생각이 듭니다. 제가 본 자료들을 몇개 올려 둘테니 천천히 확인해주세요.

- https://www.youtube.com/watch?v=g6Tg6_qpIVc
- https://www.youtube.com/watch?v=sLG5n_pXWK0
- https://www.youtube.com/watch?v=6w7SQ_1aJ0A
- https://jojoldu.tistory.com/603
- https://techblog.woowahan.com/2637/
- https://techblog.woowahan.com/12720/

## 좋았던 점
- 아키텍처, ERD 등의 프로젝트 구조에 대한 내용이 README에 담겨 있어서 프로젝트를 한눈에 파악할 수 있었습니다. 리뷰어로서 정말 행복하네요ㅎㅎ
- 아키텍처 선택에 대한 고민이 잘 드러났습니다.
- fetch join 을 활용하여 연관된 엔티티를 한번에 가져오는 쿼리를 잘 작성해주셨습니다.
- 실제 존재하는 영화 예매 시스템을 구현하기 위한 데이터 설계를 하신듯 합니다. ERD 잘해주셨어요~
- 테스트 코드도 잘 작성해주셨어요!!

## 아쉬운 점
- REST API 원칙을 조금만 더 확인해보면 어떨까요?
- API 요구사항을 다시 한번 확인해보는건 어떨까요?

## 기타 질문에 대한 답변
- 테스트 코드에 대한 영감은 아래 주소에서 많이 얻곤 했습니다ㅎㅎhttps://jojoldu.tistory.com/category/%ED%85%8C%EC%8A%A4%ED%8A%B8%EC%BD%94%EB%93%9C%20%26%20%EC%A0%95%EC%A0%81%EB%B6%84%EC%84%9D
- 개발 시간과 설계 시간은 때에 따라 달라서 뭐라 말씀 드리기가 힘들지만.. 생각보다 설계에 많은 시간이 투자됩니다. 설계에 아무 문제가 없다면 개발 시간은 얼마 안 걸리거든요.

dbdb1114 committed:
feat: 상영중인 영화 조회 수정

금일 상영 영화 조회 => 모든 상영정보 조회

dbdb1114 committed:
fix: 불필요 메소드 삭제

dbdb1114 committed:
[ refactor ] 러닝 타임 칼럼명 변경

dbdb1114 committed:
[ refactor ] 정렬 코드 제거

dbdb1114 committed:
[ refactor ] api URI 변경

restapi 원칙을 반영한 apiURI 설정

dbdb1114 committed:
fix: postgresql -> mysql

dbdb1114 committed:
[ refactor ] api 스펙 변경 반영

dbdb1114 committed:
[ feat ] 장르 조회 api 추가

dbdb1114 committed:
[ feat ] ControllerAdvice 적용

조회시 validation 추가

dbdb1114 committed:
[ refactor ] 패키지 구조 변경

dbdb1114 committed:
[ refactor ] 패키지 구조 변경

dbdb1114 committed:
[ feat ] rsd 모듈 필터링 검색 기능 추가 - entity 칼럼명 변경 및 불필요 파일 제거

entity id칼럼명 변경

dbdb1114 committed:
[ feat ] core module - response 객체 생성

dbdb1114 committed:
[ feat ] rds-repo module - 필터링 검색 기능 추가

- queryDsl 도입
- custom repository 추가 구현

dbdb1114 committed:
[ test ] rds-repo module - build.gradle 의존성 추가

dbdb1114 committed:
[ feat ] rds-repo module - 필터링 기능 테스트 코드 작성

dbdb1114 committed:
[ feat ] rsd-repo module - Genre test 코드 수정

dbdb1114 committed:
[ feat ] service module - 패키지 변경 및 필터링 검색 기능 구현

dbdb1114 committed:
[ feat ] service module - test코드 수정

dbdb1114 committed:
[ feat ] app module - 필터링 검색 기능 구현

dbdb1114 committed:
[ feat ] service module - 캐싱 기능 구현 ( 의존성 및 설정 클래스 생성 )

dbdb1114 committed:
[ feat ] service module - getTodayShowing 메소드 캐싱 적용

dbdb1114 committed:
[ feat ] redis-repo module - redis기본설정 생성

dbdb1114 committed:
[ feat ] service module - redis캐싱 기능 구현

- Caffeine 캐싱 비활성화
- redis 캐싱 정책 생성
- LocalDateTime Serialize 이슈로 인한 JacksonConfig 설정 생성

dbdb1114 committed:
Merge branch 'hanghae-skillup:main' into feature/cache

dbdb1114 committed:
[ refactor ] 불필요 interface 제거

dbdb1114 committed:
[ refactor ] 불필요 넘버링 제거

dbdb1114 committed:
[ refactor ] 400 매직넘버 제거

dbdb1114 committed:
Merge pull request #3 from dbdb1114/feature/caching

[ 2주차 완료 ] Feature/caching

dbdb1114 committed:
[ refactor ] 프로젝트 프로퍼티 통합

dbdb1114 committed:
[ test ] test 프로퍼티 Import 설정 추가

dbdb1114 committed:
[ test ] 캐싱 evict 메소드 기능 테스트

dbdb1114 committed:
Merge pull request #4 from dbdb1114/feature/caching

[ 2주차 잔여 ] Feature/caching 캐싱 테스트 완료

dbdb1114 committed:
[ refactor ] showingCache 관련 로직 제거

dbdb1114 committed:
Revert "[ refactor ] showingCache 관련 로직 제거"

This reverts commit c3163a835f6132137c52344e0e34e1f2d59678ab.

dbdb1114 committed:
[ refactor ] ShowingCacheService 생성 및 테스트 코드 수정

dbdb1114 committed:
[ refactor ] core moduel ShowingResponse stTime,edTime 칼럼명 수정

dbdb1114 committed:
[ refactor ] stTime, edTime => showStTime,showEdTime 칼럼명 변경

dbdb1114 committed:
[ fix ] redis LocalDateTime 직렬화,역직렬화 설정 수정

dbdb1114 committed:
[ fix ] 전반적인 실행환경 수정

- init.sql : 변경된 칼럼명 반영
- application-dev , docke-compose : 컴포즈 실행 환경 수정
- redisConfig : @Value 어노테이션 도입
- redis-repo/build.gradle : bootJar enabled=false  설정 추가

dbdb1114 committed:
[ fix ] 로컬에서 사용하는 클래스 ignore 설정 추가

phyeran committed:
Merge pull request #45 from dbdb1114/feature/cache

[2주차] Feature/cache

dbdb1114 committed:
[ feat ] rds-repo module - 사용자 엔티티 및 레파지토리 작성

dbdb1114 committed:
[ feat ] rds-repo module 티켓 레파지토리 및 엔티티 수정

dbdb1114 committed:
[ feat ] rds-repo module - 좌석 엔티티 및 티켓 좌석 조회 레파지토리 작성

dbdb1114 committed:
[ feat ] rds-repo module - 판매 엔티티 및 사용자별 티켓 판매 내역 정보 조회 레파지토리 작성

dbdb1114 committed:
[ feat ] rds-repo module - rating age column 추가

dbdb1114 committed:
[ fix ] rds-repo module - Movie 엔티티 getter 추가

dbdb1114 committed:
[ test ] rds-repo module 사용자별 티켓 구매 이력 조회 테스트 작성

dbdb1114 committed:
[ feat ] module-external 모듈 추가

외부 서비스 연동을 위한 모듈

dbdb1114 committed:
[ feat ]  external module - fcm 앱 푸시 서비스 작성

dbdb1114 committed:
[ feat ] module-core - 티켓 예매 서비스에 대한 커스텀 예외 정의

dbdb1114 committed:
[ feat ] service module - 티켓 조회 및 예매 기능 구현 및 테스트 코드 작성

dbdb1114 committed:
[ feat ] core module - ticket및 seates 관련 DTO 작성

dbdb1114 committed:
[ fix ] core module - spring validation 의존성 추가

dbdb1114 committed:
[ feat ] app module 예외처리 어드바이스 작성

dbdb1114 committed:
[ feat ] app module 티켓 예매 컨트롤러 및 테스트 코드 작성

dbdb1114 committed:
[ feat ] .http 작성

티켓 조회, 유저 티켓 조회, 티켓 예매 요청

dbdb1114 committed:
[ fix ] app module - 유저별 티켓 조회 수정 및 예매 시스템 기능 구현 완료

dbdb1114 committed:
[ feat ] pessimisticLock 설정 및 테스트

dbdb1114 committed:
[ feat ] OptimisticLock 설정 및 테스트

dbdb1114 committed:
Merge remote-tracking branch 'forked-origin/dev' into dev

dbdb1114 committed:
Merge branch 'dbdb1114' into dev

dbdb1114 committed:
[ Feat ] 설정 정보 변경

dbdb1114 committed:
[ feat ] redisson 기반 분산락 aop 구현

dbdb1114 committed:
[ feat ] service module aop 기반 분산락 적용

dbdb1114 committed:
[ feat ] rds-repo module 낙관적락 설정 해제

dbdb1114 committed:
[ feat ] rds-repo module @Version 제거

dbdb1114 committed:
[ feat ] 락 점유 실패 exception 정의

dbdb1114 committed:
[ feat ] service module AOP 기반 분산락 작성

AopForTransaction : 트랜잭션 분리를 위한 클래스
CustomSpringELParser :  SPEL표현식 파서
DistributedLock : 분산락 어노테이션 정의
DistributedLockAOP : aop 분산락 작성

dbdb1114 committed:
[ feat ] service module 함수형 분산락 정의

FunctionalForTransaction : 트랜잭션 분리를 위한 클래스
FunctionalDistributedLock : 함수형 분산락 구현

dbdb1114 committed:
[ feat ] service module ticketService 분산락별 메소드 분리

dbdb1114 committed:
[ feat ] service module waitTime및 leaseTime 설정

PEAK 타임을 고려한 최대 동시간대 요청 건수 계산

----------------------------------------------------------
Junghyun's hanghaeho

인기영화 100편
비인기 영화 489편
모든 영화 1일 1회씩 상영, 2일 뒤 상영정보까지 전시

----------------------------------------------------------

인기 영화 100편의 경우
상영정보 오픈과 동시에 반 정도의 티켓이 바로 판매된다고 가정
100 * 13 = 1300건

비인기 영화 489편의 경우
상영정보 오픈과 동시에 약 3매 티켓 정도 바로 판매된다고 가정
489 * 3 = 1467건

피크시간 전체 판매 티켓 수 : 2767건

----------------------------------------------------------

예매는 한 사람이 여러좌석 예매 가능하므로
피크시간 예상 요청건수 = 922.33333 건 (전체 판매 티켓 수/3)
상영정보 조회 및 선택 - 좌석 선정 프로세스 소요시간 : 최대 30초 추정

----------------------------------------------------------

테스트상 30초까지 점진적으로 1000건의 요청시 최대 요청 처리시간 : 483.43ms

- waitTime 동안 Lock을 기다릴 이유는 무엇인가?
먼저 좌석을 선정한 고객이 불가피한 이유로 결제 과정까지 넘어가지 못 했을 때

결제 과정까지 넘어가지 못하는 상황
1. 로그인, 연령 부적합 등 애초에 좌석 선정 자체가 불가한 경우
2. 결제 중 이탈
3. 단순 변심
4. ...

사실상 1번 말고는 기다렸다가 재시도 해야하고, 1번의 경우 매우 적은 수일 것으로 예상되어
실제로 Lock 점유를 기다리는 시간이 그다지 길 필요가 없다는 판단이 됩니다.
최악의 경우를 생각하여 앞의 요청 처리 두 건에서 lock점유 후 다음 프로세스로 진행이 못 했을 때 까지만 보장하기로 하여, wait_time은 1초, lease_time은 2초로 선정

lease_time 선정 이유는 만약 여러 좌석 lock 중 일부 좌석에 대한 lock을 대기하여야 할 수 있으므로 wait_time을 고려한 선정

dbdb1114 committed:
[ feat ] readme 분산락 관련 내용 정리

dbdb1114 committed:
Merge remote-tracking branch 'forked-origin/dev' into dev

dbdb1114 committed:
[ refactor ] service module lock 과정 메소드 분리

dbdb1114 committed:
[refactor] service module 함수형 분산락 락 점유 및 해제 로직  메소드 분리

dbdb1114 committed:
[ refactor ] module service reservation validate 메소드 분리

dbdb1114 committed:
[ fix ] rds module ForTransaction 클래스 private 생성자 제거

dbdb1114 committed:
[ feat ] app module RateLimit Exception 정의 및 advice 등록

dbdb1114 committed:
[ feat ] app module RateLimt 어노테이션,AOP 정의 및 적용

dbdb1114 committed:
[ feat ] app module Rate Limit RateLimiter 인터페이스 정의 및 구현체 작성

dbdb1114 committed:
[ test ] app module Rate Limit 테스트 작성

dbdb1114 committed:
[ feat ] app module redis Rate Limiter 구현

dbdb1114 committed:
[ feat ] core module CustomException 클래스 재정의

dbdb1114 committed:
[ feat ] app,service module CustomException 재정의에 따른 기존 엑셉션 사용 변경

dbdb1114 committed:
[ feat ] core, app module build.gradle 최신화

core module - SpringBootStarterWeb 의존성 추가 [ Custom Exception 정의 관련 ]
app module - aop 및 limiter 관련 의존성 추가
